Use Caffeine in `RefresingAddressResolver`

masonshin opened the pull request:
Motivation:

- Issue: #2970 Use Caffeine in `RefresingAddressResolver`
- Currently, `RefresingAddressResolver` will expire if the DNS caches is not hit. Let's keep the cache entry even if it was not hit. But it means can be increased indefinitely the size of map.
- By using Caffeine in `RefresingAddressResolver` as cache implementation, we may limit the usage even if the DNS cache entry is not expired ever.

Modifications:

- Change the life cycle of the DNS cache entry so that it does not expire.
- Use Caffeine for limiting the usage of DNS caches.

Results:
- Fixes #2970

minwoox commented on core/src/main/java/com/linecorp/armeria/common/Flags.java:
```diff
@@ -980,6 +985,20 @@ public static String fileServiceCacheSpec() {
+     * Also, specify {@code -Dcom.linecorp.armeria.dnsCache=off} JVM option to disable it.
+     */
+    @Nullable
+    public static String dnsCacheSpec() {
```
How about adding setter to `DnsResolverGroupBuilder` as well?

minwoox commented on core/src/main/java/com/linecorp/armeria/client/RefreshingAddressResolverGroup.java:
```diff
@@ -117,9 +132,15 @@
+    }
+
+    @VisibleForTesting
+    void cache(@Nullable Cache<String, CompletableFuture<CacheEntry>> dnsCache) {
```
If we add the setter to `DnsResolverGroupBuilder`, we don't need this method.
We can accept the cache spec in the constructor so that we can make `dnsCache` as a final field.

minwoox commented on core/src/main/java/com/linecorp/armeria/client/RefreshingAddressResolverGroup.java:
```diff
@@ -90,7 +90,18 @@
     }
 
-    private final ConcurrentMap<String, CompletableFuture<CacheEntry>> cache = new ConcurrentHashMap<>();
+    private static Cache<String, CompletableFuture<CacheEntry>> buildDnsCache(String spec) {
```
Could you move this private static method at the bottom of this class?
The reason that `dnsRecordTypes` is located at the top is that the method is used in the static initializer.

minwoox commented on core/src/main/java/com/linecorp/armeria/client/RefreshingAddressResolver.java:
```diff
@@ -115,30 +97,50 @@ protected void doResolve(InetSocketAddress unresolvedAddress, Promise<InetSocket
+        }
+        final String hostname = unresolvedAddress.getHostString();
+        final int port = unresolvedAddress.getPort();
+        if (cache == null) {
```
We can do this on the line 129.

minwoox commented on core/src/main/java/com/linecorp/armeria/client/RefreshingAddressResolver.java:
```diff
@@ -115,30 +97,50 @@ protected void doResolve(InetSocketAddress unresolvedAddress, Promise<InetSocket
+            return;
+        }
+        final String hostname = unresolvedAddress.getHostString();
+        final int port = unresolvedAddress.getPort();
```
Let's use the variable when it's needed. Or we can just inline this. 😄

minwoox reviewed (commented):
Basically looks good! Thank you!

ikhoon commented on core/src/main/java/com/linecorp/armeria/client/RefreshingAddressResolver.java:
```diff
@@ -65,7 +66,7 @@
 
     RefreshingAddressResolver(EventLoop eventLoop,
-                              ConcurrentMap<String, CompletableFuture<CacheEntry>> cache,
+                              @Nullable Cache<String, CompletableFuture<CacheEntry>> cache,
```
Question: Do we need `RefreshingAddressResolver` when `Cache` is null?

masonshin replied:
I was concerned about delays in updating a DNS record if someone uses DNS failover.
But it seems too much to be considered here. So I modified it to work with Nonnull.

trustin commented:
@masonshin Gentle ping 😉

masonshin commented:
Sorry for being late. I will update it soon! 🏃‍♂️

trustin commented:
@masonshin Gentle ping 😉

masonshin commented:
Sorry for being late. 🙇 
I have updated the following.
- Added `cacheSpec` setter to `DnsResolverGroupBuilder`.
- Modified `cache` is always nonnull.

minwoox commented on core/src/main/java/com/linecorp/armeria/client/DnsResolverGroupBuilder.java:
```diff
@@ -292,6 +295,18 @@ public DnsResolverGroupBuilder decodeIdn(boolean decodeIdn) {
+    /**
+     * Sets the cache spec for caching resolved addresses.
+     * If this is set, {@link Flags#dnsCacheSpec()} is ignored.
+     * Otherwise, it uses the value of {@link Flags#dnsCacheSpec()} by default.
```
I think we could just do:
```java
* Sets the cache spec for caching resolved addresses.
* {@link Flags#dnsCacheSpec()} is used by default.
```

minwoox commented on core/src/main/java/com/linecorp/armeria/client/DnsResolverGroupBuilder.java:
```diff
@@ -292,6 +295,18 @@ public DnsResolverGroupBuilder decodeIdn(boolean decodeIdn) {
+     * @see Flags#dnsCacheSpec()
+     */
+    public DnsResolverGroupBuilder cacheSpec(String cacheSpec) {
+        this.cacheSpec = cacheSpec;
```
`requireNonNull(cacheSpec, "cacheSpec");`

minwoox commented on core/src/main/java/com/linecorp/armeria/client/DnsResolverGroupBuilder.java:
```diff
@@ -343,6 +358,7 @@ RefreshingAddressResolverGroup build(EventLoopGroup eventLoopGroup) {
         return new RefreshingAddressResolverGroup(resolverConfigurator, minTtl, maxTtl, negativeTtl,
-                                                  queryTimeoutMillis, refreshBackoff, resolvedAddressTypes);
+                                                  queryTimeoutMillis, refreshBackoff, resolvedAddressTypes,
+                                                  cacheSpec);
```
How about passing `Flags#dnsCacheSpec()` is `cacheSpec` is null?

minwoox commented on core/src/main/java/com/linecorp/armeria/common/Flags.java:
```diff
@@ -355,6 +355,11 @@
+    private static final String DEFAULT_DNS_CACHE_SPEC = "maximumSize=4096";
+    @Nullable
+    private static final String DNS_CACHE_SPEC =
+            caffeineSpec("dnsCache", DEFAULT_DNS_CACHE_SPEC);
```
Let's use `dnsCacheSpec`

minwoox commented on core/src/main/java/com/linecorp/armeria/common/Flags.java:
```diff
@@ -980,6 +985,20 @@ public static String fileServiceCacheSpec() {
+     * {@link Cache} instance using {@link CaffeineSpec} for caching file entries.
+     *
+     * <p>The default value of this flag is {@value DEFAULT_DNS_CACHE_SPEC}. Specify the
+     * {@code -Dcom.linecorp.armeria.dnsCache=<spec>} JVM option to override the default value.
```
`dnsCacheSpec` here as well. 😄

ikhoon commented on core/src/main/java/com/linecorp/armeria/client/RefreshingAddressResolverGroup.java:
```diff
@@ -115,10 +116,12 @@
             dnsRecordTypes = dnsRecordTypes(resolvedAddressTypes);
         }
+        cache = buildCache(cacheSpec != null ? cacheSpec
+                                             : requireNonNull(Flags.dnsCacheSpec(), "cacheSpec"));
```
nit: Probably better to take non-null `cacheSpec` here. I think `Flags.dnsCacheSpec()` could be used before creating this class.

masonshin replied:
I've changed the `cacheSpec` as nonnull. 👍

minwoox commented on core/src/main/java/com/linecorp/armeria/client/RefreshingAddressResolverGroup.java:
```diff
@@ -137,16 +140,19 @@
+    }
+
+    private static Cache<String, CompletableFuture<CacheEntry>> buildCache(String cacheSpec) {
+        final Caffeine<Object, Object> b = Caffeine.from(cacheSpec);
```
What happens if the spec is off?

masonshin replied:
I decided to keep the DNS cache as nonnull, so I've removed `off` from the spec.

minwoox replied:
Yeah, that's a good idea. 👍

ikhoon commented on core/src/main/java/com/linecorp/armeria/client/RefreshingAddressResolver.java:
```diff
@@ -346,7 +326,6 @@ public String toString() {
                               .add("questions", questions)
                               .add("cause", cause)
                               .add("hasCacheableCause", hasCacheableCause)
-                              .add("servedFromCache", servedFromCache)
```
Not related to this PR, `numAttemptsSoFar` might be included in `toString()`.

ikhoon commented on core/src/main/java/com/linecorp/armeria/client/RefreshingAddressResolverGroup.java:
```diff
@@ -137,16 +140,19 @@
+        b.removalListener((RemovalListener<String, CompletableFuture<CacheEntry>>) (key, value, cause) -> {
+            if (value != null) {
+                value.handle((cacheEntry, throwable) -> {
                     cacheEntry.clear();
```
I guess this `RemovalListener` could be called by any thread if the `value` is completed already and `close()` is called by a user thread.
Don't we need to handle the concurrency issue in the `CacheEntry.clear()`?
For example:
```java
void clear() {
	if (executor().inEventLoop()) {
       clear0()
    } else {
       executor().execute(this::clear0)
    }
}
```

minwoox reviewed (approved):
Thanks a lot, @masonshin 👍

trustin commented on core/src/main/java/com/linecorp/armeria/client/DnsResolverGroupBuilder.java:
```diff
@@ -342,7 +354,9 @@ RefreshingAddressResolverGroup build(EventLoopGroup eventLoopGroup) {
                 builder.decodeIdn(decodeIdn);
             }
         };
+        cacheSpec = requireNonNull(cacheSpec != null ? cacheSpec : Flags.dnsCacheSpec(), "cacheSpec");
```
nit: How about declaring a final local variable rather than modifying a member field? A user did not really set the `cacheSpec`.

```suggestion
        final String cacheSpec = firstNonNull(this.cacheSpec, Flags.dnsCacheSpec());
```

trustin reviewed (approved):
Veeery nice. :+1:

trustin commented:
@masonshin Please fix the compilation error; you need to import `firstNonNull()`.

ikhoon commented on core/src/main/java/com/linecorp/armeria/client/DnsResolverGroupBuilder.java:
```diff
@@ -342,7 +355,9 @@ RefreshingAddressResolverGroup build(EventLoopGroup eventLoopGroup) {
                 builder.decodeIdn(decodeIdn);
             }
         };
+        final String cacheSpec = firstNonNull(this.cacheSpec, Flags.dnsCacheSpec());
```
`Flags.dnsCacheSpec()` is annotated with `@Nullable`. Is it intended?

masonshin replied:
It will be null when specified like: `-Dcom.linecorp.armeria.dnsCacheSpec=off`.
https://github.com/line/armeria/blob/6e0e47b60a67b8680843dae2513f051ff4f72076/core/src/main/java/com/linecorp/armeria/common/Flags.java#L1071-L1084

`Flags.dnsCacheSpec()` is itself nullable, but `firstNonNull` checks to not be null.
Would it better to make this `NonNull` when we define `dnsCacheSpec`?

ikhoon replied:
>  Would it better to make this NonNull when we define dnsCacheSpec?

+1 Because the error message of `firstNonNull(...)` would be ambiguous. Users can get a detailed error message early by making `dnsCacheSpec()` non-null.
https://github.com/google/guava/blob/7ef9f97cad23efee4cc18c63b5f3c90ca320f131/guava/src/com/google/common/base/MoreObjects.java#L64
```
"Both parameters are null"
```

masonshin replied:
SGTM 👍

ikhoon reviewed (approved):
Thanks a ton! @masonshin

trustin commented:
Thanks a lot, @masonshin. 🙇
